src/Backups/BackupIO_S3.cpp

```diff
@@ -187,6 +187,27 @@ namespace
 }
 
 
+std::shared_ptr<S3::Client> S3BackupClientFactory::getOrCreate(DiskPtr disk, S3BackupClientCreator creator)
+{
+    std::lock_guard lock(clients_mutex);
+
+    auto disk_client = disk->getS3StorageClient();
+    auto [it, inserted] = clients.try_emplace(disk->getName(), ClientEntry{nullptr, disk_client});
+    if (inserted)
+        it->second.backup_client = creator(disk);
+    else
+    {
+        auto locked = it->second.disk_reported_client.lock();
+        if (locked != disk_client)
```

Comment:
Create a new client if the underlying pointer of `disk->getS3StorageClient()` has changed, e.g., as a result of `disk->applyNewSettings`.